File: internal/application/usecase/user/create.go
package user

import (
	"context"

	"go-clean-template/internal/domain"
)

type CreateUserInput struct {
	Name     string
	Email    string
	Phone    string
	Password string
	RoleIDs  []string
}

type CreateUserOutput struct {
	User *domain.User
}

type CreateUserUseCase interface {
	Execute(ctx context.Context, input *CreateUserInput) (*CreateUserOutput, error)
}

type createUserUseCase struct {
	userRepo domain.UserRepository
	roleRepo domain.RoleRepository
}

func NewCreateUserUseCase(userRepo domain.UserRepository, roleRepo domain.RoleRepository) CreateUserUseCase {
	return &createUserUseCase{userRepo: userRepo, roleRepo: roleRepo}
}

func (uc *createUserUseCase) Execute(ctx context.Context, input *CreateUserInput) (*CreateUserOutput, error) {
	user, err := domain.NewUser(input.Name, input.Email, input.Phone, input.RoleIDs)
	if err != nil {
		return nil, err
	}
	if err := user.SetPassword(input.Password); err != nil {
		return nil, err
	}
	existing, _ := uc.userRepo.FindByEmail(ctx, input.Email)
	if existing != nil {
		return nil, domain.ErrAlreadyExists
	}
	if err := validateRoleIDs(ctx, uc.roleRepo, input.RoleIDs); err != nil {
		return nil, err
	}
	if err := uc.userRepo.Create(ctx, user); err != nil {
		return nil, err
	}
	return &CreateUserOutput{User: user}, nil
}

File: src/application/usecase/user/create.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::{RoleRepository, UserRepository};
use crate::domain;

pub struct CreateUserInput {
    pub name: String,
    pub email: String,
    pub phone: String,
    pub password: String,
    pub role_ids: Vec<String>,
}

pub struct CreateUserOutput {
    pub user: domain::User,
}

#[async_trait]
pub trait CreateUserUseCase: Send + Sync {
    async fn execute(&self, input: CreateUserInput) -> Result<CreateUserOutput, AppError>;
}

pub struct CreateUserUseCaseImpl {
    user_repo: Arc<dyn UserRepository>,
    role_repo: Arc<dyn RoleRepository>,
}

impl CreateUserUseCaseImpl {
    pub fn new(user_repo: Arc<dyn UserRepository>, role_repo: Arc<dyn RoleRepository>) -> Self {
        CreateUserUseCaseImpl {
            user_repo,
            role_repo,
        }
    }
}

#[async_trait]
impl CreateUserUseCase for CreateUserUseCaseImpl {
    async fn execute(&self, input: CreateUserInput) -> Result<CreateUserOutput, AppError> {
        let mut user = domain::User::new(input.name, input.email, input.phone, input.role_ids)?;
        user.set_password(&input.password)?;

        let existing = self.user_repo.find_by_email(&user.email).await;
        if existing.is_ok() {
            return Err(AppError::from(domain::Error::AlreadyExists));
        }

        super::helpers::validate_role_ids(&*self.role_repo, &user.role_ids).await?;

        self.user_repo.create(&user).await?;
        Ok(CreateUserOutput { user })
    }
}
